src/cswinrt/code_writers.h

```diff
@@ -1805,87 +1808,89 @@ remove => %;
+            (get<uint16_t>(get_arg(2)) >> 8) & 0xFF,
+            get<uint8_t>(get_arg(3)),
+            get<uint8_t>(get_arg(4)),
+            get<uint8_t>(get_arg(5)),
+            get<uint8_t>(get_arg(6)),
+            get<uint8_t>(get_arg(7)),
+            get<uint8_t>(get_arg(8)),
+            get<uint8_t>(get_arg(9)),
+            get<uint8_t>(get_arg(10)));
+    }
+
+    void write_class_static_cache_definition(writer& w, TypeDef const& staticsType, TypeDef const& classType)
     {
-        auto instance =
-            w.write_temp(
-                "private static readonly _% _instance = new _%();",
-                cache_type_name,
-                cache_type_name);
-        auto factoryAs = w.write_temp("%",
-            bind([&](writer& w)
-            {
-                if (is_static(class_type))
-                {
-                    w.write("_factory._As");
-                }
-                else
-                {
-                    w.write("ActivationFactory<%>.As", class_type.TypeName());
-                }
-            })
-        );
+        if (staticsType)
+        {
+            auto factory_class_name = settings.netstandard_compat ?
+                w.write_temp("BaseFactory<%.Vftbl>", bind<write_type_name>(staticsType, typedef_name_type::ABI, true)) :
+                w.write_temp("BaseFactory");
 
-        if (settings.netstandard_compat)
-        {            
-        auto cache_vftbl_type = w.write_temp("ABI.%.%.Vftbl",
-                class_type.TypeNamespace(),
-                cache_type_name);
-        auto cache_interface =
-            w.write_temp(
-                R"(%<%>)",
-                factoryAs,
-                cache_vftbl_type);
+            auto statics_type_name = staticsType.TypeName();
             w.write(R"(
-internal sealed class _% : ABI.%.%
-{
-public _%() : base(%()) { }
-%
-internal static % Instance => _instance;
-}
+private static % _% = new %("%.%", %.IID);
 )",
-                cache_type_name,
-                class_type.TypeNamespace(),
-                cache_type_name,
-                cache_type_name,
-                cache_interface,
-                instance,
-                cache_type_name);
+                factory_class_name,
+                statics_type_name,
+                factory_class_name,
+                classType.TypeNamespace(),
+                classType.TypeName(),
+                bind<write_type_name>(staticsType, typedef_name_type::StaticAbiClass, true));
         }
-        else
-        {
-            w.write(R"(
-internal sealed class _%
-{
-private IObjectReference _obj;
-public _%()
-{
-_obj = %(%.IID);
-}
+    }
 
-%
-internal static % Instance => (%)_instance;
-}
+    void write_activation_factory_static_cache_definition(writer& w, TypeDef const& classType)
+    {
+        auto class_type_name = classType.TypeName();
+        w.write(R"(
+private static BaseActivationFactory _%Factory = new BaseActivationFactory("%.%");
```

Comment:
Same for this one, shouldn't it be `readonly`, or could this ever be reassigned?